[NPUW] Add Eagle3 (top-1) pipeline support with new I/O

GuoliangShiIntel opened the pull request:
### Details:
This PR introduces support for Eagle3 (top-1) speculative decoding in the NPUW. The main changes include:

1. Added a new `llm_eagle3_extension` module to handle Eagle3-specific input/output logic, including model role detection (Draft/Target), input padding, and chunked processing.
2. Updated `LLMInferRequest` to automatically detect Eagle3 models and manage Eagle3 input/output tensors during prefill and generate stages.
3. Modified model reshaping and output redirection functions to support new Eagle3 layer names and shapes.

These changes enable integration of Eagle3 (top-1) speculative decoding models with the NPU plugin.

### Tickets:
 - [*CVS-175433*](https://jira.devtools.intel.com/browse/CVS-175433)

<img width="2472" height="970" alt="image" src="https://github.com/user-attachments/assets/9c3b74ae-7346-492e-85bb-4ac569ae0d09" />

AsyaPronina commented on src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.hpp:
```diff
@@ -102,6 +102,8 @@ class LLMCompiledModel : public ov::npuw::ICompiledModel {
 
     bool m_is_whisper = false;
+
+    ov::AnyMap m_model_rt_info;
```
We are now considering all newly added fields to the `LLMCompiledModel` as the fields that participate in serialization/deserialization in export/import and caching activities. However, I am not sure about this field, it needs to be checked how to handle it for import and caching.

AsyaPronina commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,197 @@
+            LOG_VERB("Eagle3 " << (m_role == Eagle3ModelRole::Draft ? "Draft" : "Target")
+                               << ": Retrieved last_hidden_state output tensor");
+        }
+    }
```
This function doesn't seem to do any type of processing for output, it just retrieves and stores it. Might be it can be named something like `update_last_hidden_state()` or something like this.

GuoliangShiIntel commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,197 @@
+                uint32_t total_tokens = static_cast<uint32_t>(original_shape[seq_dim]);
+                uint32_t chunk_end_token = std::min(chunk_start_token + chunk_token_count, total_tokens);
+
+                if (chunk_start_token < total_tokens && chunk_token_count > 0) {
```
We shouldn't hit this scenario. I've replaced it with a static assert to ensure the chunk stays within bounds.

GuoliangShiIntel commented on src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:
```diff
@@ -528,8 +530,8 @@ void ov::npuw::LLMInferRequest::trim_kvcache_for_speculative_decoding(ov::SoPtr<
     if (dirty_num > 0) {
-        LOG_DEBUG("Trim kv cache from " << kvcache_desc.num_stored_tokens << " length"
-                                        << " to " << position_id << " length");
+        LOG_DEBUG("Trim kv cache from " << kvcache_desc.num_stored_tokens << " length" << " to " << position_id
```
Yes, it is, change back

GuoliangShiIntel commented on src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:
```diff
@@ -895,6 +920,11 @@ void ov::npuw::LLMInferRequest::infer() {
     OPENVINO_ASSERT(ov::element::i64 == position_ids->get_element_type());
 
+    // Eagle3 Draft: Process user-provided hidden state inputs for draft model
+    m_eagle3_ext.store_user_tensors(inputs, [this](const ov::Output<const ov::Node>& port) {
```
I haven't changed the current approach here. I think it's better to keep all Eagle-related logic contained within Eagle3Extension to minimize impact on the inference pipeline.

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:
```diff
@@ -1908,6 +1924,7 @@ void ov::npuw::LLMCompiledModel::serialize(std::ostream& stream, const ov::npuw:
         write(model_stream, m_prefix_caching_max_num_blocks);
         write(model_stream, m_gemma_sliding_window_size);
         write(model_stream, m_is_whisper);
+        write(model_stream, m_model_rt_info);
```
I don't think this is a good idea.
rt info could contain a lot of information which is not directly affects compilation or even have no sense for deserialized model.

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,246 @@
+    }
+}
+
+void Eagle3Extension::prepare_inputs(std::shared_ptr<ov::IAsyncInferRequest> request,
```
```c++
const std::shared_ptr<ov::IAsyncInferRequest>& request
```

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,246 @@
+}
+
+void Eagle3Extension::update_last_hidden_state(
+    std::shared_ptr<ov::IAsyncInferRequest> request,
```
```c++
const std::shared_ptr<ov::IAsyncInferRequest>& request
```

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.hpp:
```diff
@@ -0,0 +1,104 @@
+        const std::function<ov::SoPtr<ov::ITensor>(const ov::Output<const ov::Node>&)>& get_tensor_func);
+
+    // Prepare Eagle3 input tensors (hidden_states, internal_hidden_states) for draft models
+    void prepare_inputs(std::shared_ptr<ov::IAsyncInferRequest> request,
```
```c++
const std::shared_ptr<ov::IAsyncInferRequest>& request
```

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:
```diff
@@ -1120,10 +1143,26 @@ void ov::npuw::LLMInferRequest::infer() {
     }
+
+    if (port_names.count(Eagle3LayerNames::last_hidden_state) > 0) {
+        if (!m_eagle3_ext.is_eagle3_model()) {
```
I'd better swap conditions:
```c++
if (m_eagle3_ext.is_eagle3_model()) {
    if (port_names.count(Eagle3LayerNames::last_hidden_state) > 0) {
        ...
}
```

GuoliangShiIntel commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,246 @@
+    OPENVINO_ASSERT(padded_shape[1] >= hidden_state_shape[1], "Padded token length must be >= input token length");
+
+    // Zero-fill the padded tensor
+    std::memset(padded_hidden_state->data(), 0, padded_hidden_state->get_byte_size());
```
Done

AlexanderKalistratov commented on src/plugins/intel_npu/src/plugin/npuw/llm_compiled_model.cpp:
```diff
@@ -2132,6 +2168,7 @@ std::shared_ptr<ov::npuw::LLMCompiledModel> ov::npuw::LLMCompiledModel::deserial
         read(model_stream, compiled->m_prefix_caching_max_num_blocks);
         read(model_stream, compiled->m_gemma_sliding_window_size);
         read(model_stream, compiled->m_is_whisper);
+        read(model_stream, compiled->m_model_rt_info);
```
Serializing flag only wouldn't solve the issue.
Please do not change this code until we'll figure out how to handle it properly.
It is probably should be property, not flag set via `rt_info`

GuoliangShiIntel commented:
@AsyaPronina @AlexanderKalistratov 

Today, I aligned with the GPU pipeline and identified some model structure changes aimed at improving performance. Consequently, we will implement similar changes for the NPU pipeline, as we will be sharing the same model transformation.

**The change are:**
1. Moving the fully convolutional layer from the draft model to the target model
The target model will concatenate three hidden states along the last dimension and then use a fully connected layer to convert it back to the original shape.
Previously, the target output shape was `[1, token_len, 3*embedding_size]`. Now, the output shape is `[1, token_len, embedding_size]`.

2. Remove the `internal_hidden_states` input from the draft model
After the first change, the `hidden_states` input has the same shape as `internal_hidden_states`. Additionally, since only one of these inputs is activated at any stage while the other remains zero and they are summed together, we can maintain a single input and control which one is passed in through the pipeline.

The change in npuw with this [commit](https://github.com/openvinotoolkit/openvino/pull/32494/commits/d5468c4db923b388b88ba82128bd08929fef473a)

AsyaPronina commented on src/plugins/intel_npu/src/plugin/npuw/llm_eagle3_extension.cpp:
```diff
@@ -0,0 +1,211 @@
+
+void Eagle3Extension::store_hidden_state_inputs(
+    const std::vector<ov::Output<const ov::Node>>& inputs,
+    const std::function<ov::SoPtr<ov::ITensor>(const ov::Output<const ov::Node>&)>& get_tensor_func) {
```
Could we just use link to the infer request here?

AsyaPronina commented on src/plugins/intel_npu/src/plugin/npuw/llm_infer_request.cpp:
```diff
@@ -1072,6 +1090,11 @@ void ov::npuw::LLMInferRequest::infer() {
     OPENVINO_ASSERT(ov::element::i64 == position_ids->get_element_type());
 
+    // Eagle3: Accept and validate hidden state inputs
+    m_eagle3_ext.store_hidden_state_inputs(inputs, [this](const ov::Output<const ov::Node>& port) {
```
As we have one input now, getter seems a bit redundant, lets stick to the `infer_request` and input ports map here

AsyaPronina reviewed (commented):
Hello @GuoliangShiIntel !
Thanks a lot for a great work!
I am ready to approve, but could I please ask you first to reflect new property also for caching:
https://github.com/intelgaoxiong/openvino/blob/3bc0ca5ee8b79456c0da71b7578a45157e196838/src/plugins/intel_npu/src/plugin/include/properties.hpp#L69

GuoliangShiIntel commented:
Hi @AsyaPronina @AlexanderKalistratov I've addressed all review comments. What's the next step? Do we need additional reviews, or should we wait for the GenAI Pipeline PR to be ready before merging?

AlexanderKalistratov reviewed (approved):
Thanks a lot for the PR!

AsyaPronina commented:
build_jenkins

AsyaPronina commented:
Hello! Validation results are all successful! PR can be merged!
However, I would like to ask you to make a follow-up PR where `m_eagle3_ext.store_hidden_state_inputs(*this, inputs);` in `llm_infer_request.cpp` will be also wrapped under `if (m_eagle3_ext.is_eagle3_model())` condition. It is not required per code logic, but would be great to see it aligned with other usages of extension!
